Code/Tools/AssetProcessor/native/utilities/Builder.h

```diff
@@ -0,0 +1,165 @@
+        bool IsValid() const;
+
+        //! Returns true if the builder has a process watcher and the process is running OR does not have a process watcher.  Returns false
+        //! otherwise
+        bool IsRunning(AZ::u32* exitCode = nullptr) const;
+
+        //! Returns true if the builder exe has established a connection
+        bool IsConnected() const;
+
+        //! Blocks waiting for the builder to establish a connection
+        AZ::Outcome<void, AZStd::string> WaitForConnection();
+
+        AZ::u32 GetConnectionId() const;
+        AZ::Uuid GetUuid() const;
+        AZStd::string UuidString() const;
+
+        void PumpCommunicator() const;
+        void FlushCommunicator() const;
+        void TerminateProcess(AZ::u32 exitCode) const;
+
+        //! Sends the job over to the builder and blocks until the response is received or the builder crashes/times out
+        template<typename TNetRequest, typename TNetResponse, typename TRequest, typename TResponse>
+        BuilderRunJobOutcome RunJob(
+            const TRequest& request,
+            TResponse& response,
+            AZ::u32 processTimeoutLimitInSeconds,
+            const AZStd::string& task,
+            const AZStd::string& modulePath,
+            AssetBuilderSDK::JobCancelListener* jobCancelListener = nullptr,
+            AZStd::string tempFolderPath = AZStd::string()) const;
+
+    protected:
+        //! Starts the builder process and waits for it to connect
+        virtual AZ::Outcome<void, AZStd::string> Start(BuilderPurpose purpose);
+
+        //! Sets the connection id and signals that the builder has connected
+        void SetConnection(AZ::u32 connId);
+
+        AZStd::vector<AZStd::string> BuildParams(
+            const char* task,
+            const char* moduleFilePath,
+            const AZStd::string& builderGuid,
+            const AZStd::string& jobDescriptionFile,
+            const AZStd::string& jobResponseFile,
+            BuilderPurpose purpose) const;
+        AZStd::unique_ptr<AzFramework::ProcessWatcher> LaunchProcess(
+            const char* fullExePath, const AZStd::vector<AZStd::string>& params) const;
+
+        //! Waits for the builder exe to send the job response and pumps stdout/err
+        BuilderRunJobOutcome WaitForBuilderResponse(
+            AssetBuilderSDK::JobCancelListener* jobCancelListener,
+            AZ::u32 processTimeoutLimitInSeconds,
+            AZStd::binary_semaphore* waitEvent) const;
+
+        //! Writes the request out to disk for debug purposes and logs info on how to manually run the asset builder
+        template<typename TRequest>
+        bool DebugWriteRequestFile(
+            QString tempFolderPath,
+            const TRequest& request,
+            const AZStd::string& task,
+            const AZStd::string& modulePath,
+            BuilderPurpose purpose) const;
+
+        const AZ::Uuid m_uuid;
```

Comment:
Reason for all these to be protected? Does every inheriting builder need access?

Reply:
Mostly for unit tests, to avoid having to reorganize things around from the private block to the protected block every time a function/member is needed.  Unit tests are the only place this class is overridden.